mountpoint-s3-crt/src/io/async_stream.rs

```diff
@@ -0,0 +1,246 @@
+        "this function should only be called on streams that use the impl vtable"
+    );
+
+    assert!(!(*stream).impl_.is_null(), "stream.impl_ should never be null");
+
+    // SAFETY: On streams using the impl vtable, .impl_ always is a ptr to a
+    // AsyncInputStreamImpl.
+    let impl_ptr = (*stream).impl_ as *mut AsyncInputStreamImpl;
+
+    assert!(
+        std::ptr::eq(&(*impl_ptr).inner, stream),
+        "&async_input_stream.inner should be the same stream we started with"
+    );
+
+    impl_ptr
+}
+
+unsafe extern "C" fn destroy_impl(stream: *mut aws_async_input_stream) {
+    let impl_stream = async_input_stream_to_impl(stream);
+
+    // SAFETY: This is always a Box pointer, since we know this is from the impl vtable.
+    // This will `drop` the stream's contents if it's the last pointer.
+    drop(Box::from_raw(impl_stream));
+}
+
+unsafe extern "C" fn read_impl(stream: *mut aws_async_input_stream, dest: *mut aws_byte_buf) -> *mut aws_future_bool {
+    let stream = &mut *async_input_stream_to_impl(stream);
+
+    let promise = BoolPromise::new(stream.inner.alloc);
+    let request = ReadRequest::new(dest, promise.clone());
+    if let Err(e) = stream.sender.try_send(request) {
+        debug!("channel closed: {:?}", e);
+        e.into_inner().complete();
+    }
+
+    promise.leak()
+}
+
+/// Fulfills a [aws_future_bool].
+#[derive(Debug)]
+struct BoolPromise {
+    inner: NonNull<aws_future_bool>,
+}
+
+impl BoolPromise {
+    fn new(alloc: *mut aws_allocator) -> Self {
+        Self {
+            // SAFETY: aws_future_bool_new returns a valid aws_future_bool.
+            inner: unsafe { NonNull::new_unchecked(aws_future_bool_new(alloc)) },
+        }
+    }
+
+    fn fulfill(self, result: Result<bool, Error>) {
+        match result {
+            // SAFETY: inner contains a valid aws_future_bool.
+            Ok(value) => unsafe {
+                aws_future_bool_set_result(self.inner.as_ptr(), value);
+            },
+            // SAFETY: inner contains a valid aws_future_bool.
+            Err(err) => unsafe {
+                aws_future_bool_set_error(self.inner.as_ptr(), err.raw_error());
+            },
+        }
+    }
+
+    fn leak(self) -> *mut aws_future_bool {
+        // SAFETY: inner contains a valid aws_future_bool.
+        unsafe { aws_future_bool_acquire(self.inner.as_ptr()) }
+    }
+}
+
+// SAFETY: aws_future_bool is atomically ref-counted and its methods are thread-safe.
+unsafe impl Send for BoolPromise {}
+
+impl Drop for BoolPromise {
+    fn drop(&mut self) {
+        // SAFETY: inner contains a valid aws_future_bool.
+        unsafe {
+            aws_future_bool_release(self.inner.as_ptr());
+        }
+    }
+}
+
+impl Clone for BoolPromise {
+    fn clone(&self) -> Self {
+        // SAFETY: inner contains a valid aws_future_bool.
+        let inner = unsafe { NonNull::new_unchecked(aws_future_bool_acquire(self.inner.as_ptr())) };
+        Self { inner }
+    }
+}
```

Comment:
Think it's worth trying to write a few tests specifically for this code. Maybe spawn two threads and have one read while the other writes, or something. Try to exercise the various failure cases (what if the reader gives up? what if the writer gives up?).